sys/src/9/port/devramfs.c

```diff
@@ -0,0 +1,180 @@
+/*
+ * This file is part of the UCB release of Plan 9. It is subject to the license
+ * terms in the LICENSE file found in the top-level directory of this
+ * distribution and at http://akaros.cs.berkeley.edu/files/Plan9License. No
+ * part of the UCB release of Plan 9, including this file, may be copied,
+ * modified, propagated, or distributed except according to the terms contained
+ * in the LICENSE file.
+ */
+
+#include	"u.h"
+#include	"../port/lib.h"
+#include	"mem.h"
+#include	"dat.h"
+#include	"fns.h"
+#include	"../port/error.h"
+
+#define DIR_EXTEND_SIZE 10
+
+int memcpy(void* dst, const void* src, int len);
+
+struct Directory {
+	int inuse;
+	int allocated;
+	Dirtab ramdir[];
+};
+
+struct RamFile {
+	int length;
+	char data[];
+};
+
+static struct Directory *ramroot;
+
+static char* readme = "This is the kernel ramdisk.\n";
+
+static void
+raminit(void)
+{
+	ramroot = (struct Directory *)smalloc(sizeof(struct Directory) + DIR_EXTEND_SIZE * sizeof(Dirtab));
+	ramroot->inuse = 3;
+	ramroot->allocated = DIR_EXTEND_SIZE;
+	strcpy(ramroot->ramdir[0].name, ".");
+	ramroot->ramdir[0].qid = (Qid){(int64_t)ramroot, 0,QTDIR};
+	ramroot->ramdir[0].length = 0;
+	ramroot->ramdir[0].perm = DMDIR|0777;
+
+	struct RamFile* file = (struct RamFile*)smalloc(sizeof(struct RamFile*) + strlen(readme));
+	file->length = strlen(readme);
+	memcpy(file->data, readme, file->length);
+	strcpy(ramroot->ramdir[1].name, "README");
+	ramroot->ramdir[1].qid = (Qid){(int64_t)file};
+	ramroot->ramdir[1].length = file->length;
+	ramroot->ramdir[1].perm = 0444;
+
+	struct Directory *subdir = (struct Directory *)smalloc(sizeof(struct Directory) + DIR_EXTEND_SIZE * sizeof(Dirtab));
+	strcpy(ramroot->ramdir[2].name, "subdir");
+	ramroot->ramdir[2].qid = (Qid){(int64_t)subdir, 0, QTDIR};
+	ramroot->ramdir[2].length = 0;
+        ramroot->ramdir[2].perm = DMDIR|0777;
+
+	subdir->inuse = 3;
+        subdir->allocated = DIR_EXTEND_SIZE;
+        strcpy(subdir->ramdir[0].name, "..");
+        subdir->ramdir[0].qid = (Qid){(int64_t)ramroot, 0,QTDIR};
+        subdir->ramdir[0].length = 0;
+        subdir->ramdir[0].perm = DMDIR|0777;
+
+        strcpy(subdir->ramdir[1].name, ".");
+        subdir->ramdir[1].qid = (Qid){(int64_t)subdir, 0,QTDIR};
+        subdir->ramdir[1].length = 0;
+        subdir->ramdir[1].perm = DMDIR|0777;
+
+
+        strcpy(subdir->ramdir[2].name, "README2");
+        subdir->ramdir[2].qid = (Qid){(int64_t)file};
+        subdir->ramdir[2].length = file->length;
+        subdir->ramdir[2].perm = 0444;
+
+}
+
+static void
+ramreset(void)
+{
+}
+
+static Chan*
+ramattach(char *spec)
+{
+        Chan *c;
+        char *buf;
+         
+        c = newchan();
+        mkqid(&c->qid, (int64_t)ramroot, 0, QTDIR);
+        c->dev = devtabget('@', 0);
+        if(spec == nil)
+                spec = "";
+        buf = smalloc(1+UTFmax+strlen(spec)+1);
+        sprint(buf, "#@%s", spec);
+        c->path = newpath(buf);
+	c->mode = 0777;
```

Comment:
you might want to run clang-format on this to get indentation and such things right.